### PR TITLE
[BUGFIX] Fluent Datasources - empty "assets" key on serialization

### PR DESCRIPTION
**Pull request by Kilo59**
## Changes proposed in this pull request:
- Remove `'assets'` from `_FIELDS_ALWAYS_SET`
- Only consider the `assets` part of `__fields_set__` if the `Datasource` actually contains `assets`.
- Add private function `_update__fields_set__on_truthyness()`

### Definition of Done

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

**Comment by ghost** [comment]
<details open>
<summary><b>👇 <a href="https://app.codesee.io/r/reviews?pr=7341&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations">Click on the image</a> for a new way to code review</b>
</summary>

#### 

[Review these changes using an interactive CodeSee Map](https://app.codesee.io/r/reviews?pr=7341&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations)

#### Legend
CodeSee Map legend
</details>

File: tests/datasource/fluent/test_config.py
Comment (Kilo59) [inline tests/datasource/fluent/test_config.py:503]:
```suggestion
    assert ds_with_assets, "Need at least one Datasource with assets for this test"
    assert ds_wo_assets, "Need at least one Datasource without assets for this test"
```

**Comment by Kilo59** [comment]
Wait for merge from
- https://github.com/great-expectations/great_expectations/pull/7336

File: great_expectations/datasource/fluent/interfaces.py
Comment (NathanFarmer) [inline great_expectations/datasource/fluent/interfaces.py:468]: Why don't we need this anymore?

File: tests/datasource/fluent/test_metadatasource.py
Comment (NathanFarmer) [inline tests/datasource/fluent/test_metadatasource.py:412]: ❤️

**Review by NathanFarmer — APPROVED** [review]

File: great_expectations/datasource/fluent/interfaces.py
Comment (Kilo59) [reply great_expectations/datasource/fluent/interfaces.py:468]: Because we do this same logic before serialization. It's very likely we didn't actually need this before.

File: great_expectations/datasource/fluent/fluent_base_model.py
Comment (alexsherstinsky) [inline great_expectations/datasource/fluent/fluent_base_model.py:306]: @Kilo59 Do we need the double underscore between `set` and `on` in this method's name?  Thanks.

File: great_expectations/datasource/fluent/constants.py
Comment (alexsherstinsky) [inline great_expectations/datasource/fluent/constants.py:8]: ❤️

File: great_expectations/datasource/fluent/fluent_base_model.py
Comment (alexsherstinsky) [inline great_expectations/datasource/fluent/fluent_base_model.py:307]: @Kilo59 I am approving, because I do not have a better proposal at the moment, other than to express that being overly invested in the inner workings of a framework (`Pydantic` in this case) may cause maintenance difficulties in the future (e.g., if the framework's behavior changes).  In addition, the use of `getattr()` masks the specifics of the the field being used, and while appropriate in certain situation may cause unforeseen results, if used on unintended field names (e.g., by accident) -- @cdkini may wish to offer an opinion on `getattr()`.  Finally, a question: is the absence of `assets` a normal situation, or a corner case?  If the former, then perhaps the `debug` log messages may not be needed.  Thank you.

File: tests/datasource/fluent/test_config.py
Comment (alexsherstinsky) [inline tests/datasource/fluent/test_config.py]: @Kilo59 Non-blocking: stylistically, when the `if` clause contains the `else`, it might be less verbose to put the negative branch into the `else` part.  Thanks.

File: tests/datasource/fluent/test_config.py
Comment (alexsherstinsky) [inline tests/datasource/fluent/test_config.py:587]: @Kilo59 Nice way to test all variations without additional fixtures!

**Review by alexsherstinsky — CHANGES_REQUESTED** [review]
Will LGTM (after resolving the method name with the double underscore and acknowledging the non-blocking comments).  Thanks.

File: great_expectations/datasource/fluent/fluent_base_model.py
Comment (Kilo59) [reply great_expectations/datasource/fluent/fluent_base_model.py:306]: It is referencing the `__fields_set__` attribute of a pydantic model.

File: great_expectations/datasource/fluent/fluent_base_model.py
Comment (Kilo59) [reply great_expectations/datasource/fluent/fluent_base_model.py:307]: Yes, we are relying on some implementation details for `pydantic` here. But all of the things we rely on are part of it's public documentation (example `__fields_set__`- https://docs.pydantic.dev/usage/models/#basic-model-usage )
I'm not worried about the framework changing because we are pinned to the last major version. If `pydantic` makes breaking changes (for us) in `v2` we don't have to use it.

>In addition, the use of getattr() masks the specifics of the the field being used, and while appropriate in certain situation may cause unforeseen results, if used on unintended field names (e.g., by accident)

I don't follow.
It is true that this logic doesn't need to be applied to `FluentBaseModel` and could be isolated to `Datasource.dict()`/`Datasource.json()`. But this is no worse than what we had previously and I wanted to keep these changes small and address the bug.

Comment (Kilo59) [reply great_expectations/datasource/fluent/fluent_base_model.py:307]: >Finally, a question: is the absence of assets a normal situation, or a corner case?

I would say it's very normal. A Datasource can be created and then saved or serialized without any `assets`.

File: tests/datasource/fluent/test_config.py
Comment (Kilo59) [inline tests/datasource/fluent/test_config.py:501]:
```suggestion
        if ds.assets:
            ds_with_assets = ds
        else:
            ds_wo_assets = ds
```

File: tests/datasource/fluent/test_config.py
Comment (Kilo59) [reply tests/datasource/fluent/test_config.py]: I agree.  Updated.

File: great_expectations/datasource/fluent/fluent_base_model.py
Comment (alexsherstinsky) [reply great_expectations/datasource/fluent/fluent_base_model.py:307]: Hence, the log messages are not needed perhaps?

Comment (Kilo59) [reply great_expectations/datasource/fluent/fluent_base_model.py:307]: I would prefer to keep the debug logs, at least for now.
We log at `warning` and above, users won't see this.

File: tests/datasource/fluent/test_config.py
Comment (alexsherstinsky) [reply tests/datasource/fluent/test_config.py]: 🙇

File: great_expectations/datasource/fluent/fluent_base_model.py
Comment (alexsherstinsky) [reply great_expectations/datasource/fluent/fluent_base_model.py:306]: @Kilo59 Good point -- thank you.

**Review by alexsherstinsky — APPROVED** [review]
LGTM

File: tests/datasource/fluent/test_config.py
Comment (Kilo59) [inline tests/datasource/fluent/test_config.py:763]:
```suggestion
```
This should now give us a key error and needs to be removed.

**Pull request merged**
